vllm/model_executor/models/minicpmv.py

```diff
@@ -568,110 +509,60 @@ def get_image_bounds(self, input_ids: torch.Tensor):
-
-                    vision_hidden_states = self.resampler(
-                        vision_embedding, tgt_sizes)
-
-                else:  # no image
-                    dummy_feature = []
-                    vision_hidden_states = dummy_feature
+    def get_embedding(
+        self,
+        input_ids: torch.Tensor,
+        image_inputs: Optional[MiniCPMVImageInputs],
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        if image_inputs is None:  # No image
+            vlm_embedding = torch.tensor([], device=input_ids.device)
+            vision_hidden_states = torch.tensor([], device=input_ids.device)
         else:
-            vision_hidden_states = data["vision_hidden_states"]
-
-        return vision_hidden_states
-
-    def get_embedding(self, data: Dict[str, Union[List[torch.Tensor],
-                                                  torch.Tensor]]):
-        input_ids = data["input_ids"]
-
-        vision_hidden_states = self.get_vision_hidden_states(data)
-        if vision_hidden_states is not None and len(vision_hidden_states) > 0:
+            vision_hidden_states = self.get_vision_hidden_states(image_inputs)
             image_bounds = self.get_image_bounds(input_ids)
-        else:
-            image_bounds = []
 
-        if hasattr(self.config, 'scale_emb'):
-            vlm_embedding = self.llm.embed_tokens(
-                input_ids) * self.config.scale_emb
-        else:
-            vlm_embedding = self.llm.embed_tokens(input_ids)
-        vision_hidden_states = [
-            i.type(vlm_embedding.dtype) if isinstance(i, torch.Tensor) else i
-            for i in vision_hidden_states
-        ]
+            if hasattr(self.config, "scale_emb"):
+                vlm_embedding: torch.Tensor = (
+                    self.llm.embed_tokens(input_ids) * self.config.scale_emb)
+            else:
+                vlm_embedding = self.llm.embed_tokens(input_ids)
 
-        if len(vision_hidden_states) > 0 and len(image_bounds) > 0:
-            vision_hidden_states = torch.cat(vision_hidden_states, dim=0)
             image_indices = torch.stack([
                 torch.arange(r[0], r[1], dtype=torch.long)
                 for r in image_bounds
-            ]).to(vlm_embedding.device)
+            ]).to(vlm_embedding.device, vlm_embedding.dtype)
             vlm_embedding.scatter_(
                 0,
                 image_indices.view(-1, 1).repeat(1, vlm_embedding.shape[-1]),
-                vision_hidden_states.view(-1, vision_hidden_states.shape[-1]))
+                vision_hidden_states.view(-1, vision_hidden_states.shape[-1]),
+            )
+
         return vlm_embedding, vision_hidden_states
 
-    def process_multimodal_inputs(self, inputs: Dict[str,
-                                                     Union[List[torch.Tensor],
-                                                           torch.Tensor]]):
-        pixel_values = []
-        tgt_sizes = []
-        for b in range(len(inputs["pixel_values"])):
```

Comment:
This code seems missing.

Reply:
I intentionally avoided converting them into a list as the code seems to support both. Is there a place where this isn't true?

Reply:
The thing is that, as input images contains a batch dimension with the shape of `[B, N, 3, H, W]`. While the `input_ids` is squeezed to 1-D, the images should be so. So I added this.

Reply:
I think I made a mistake about here. I realize you've already do this in `try_concat` funciton.

Reply:
Oh, I get it now, you're basically flattening the 2-D nested list into a 1-D one. Fixing...

Reply:
I implemented the `image_processor` in a slightly different way compared with `LLava-Next`. ``LLava-Next` put all the images together and output them together in `processor` and `image_processor`, then left them split in the model. But the original code of `MiniCPMV` doesn't do this, so in my implementation, there's an extra `batch` dimension.

Reply:
Fixed.